### PR TITLE
Support sum(sparse_csr)

### PR DESCRIPTION
**Pull request by cpuhrsch**
Basic support for summation of CSR. ~~Generalizes structured torch.sum to also support CSR.~~

Follow up work:
- Autograd support
- OpInfo integration

### CODE REVIEW

File: tools/codegen/dest/register_dispatch_key.py [hunk 1]
```diff
@@ -86,11 +86,13 @@ def gen_create_out_helper(backend_index: BackendIndex) -> List[str]:
     return [f"""
 Tensor create_out(IntArrayRef sizes, IntArrayRef strides, const TensorOptions &options) {{
   {runtime_empty_supported_check}
+  if ({empty_options}.layout() == c10::kSparseCsr && sizes.size() == 0) {{
+      return {empty_impl}(sizes, {empty_options}.layout(c10::kStrided));
```
Comment (ezyang) [inline tools/codegen/dest/register_dispatch_key.py:90]: Does this always make sense? It seems unlikely that if you have a CSR tensor you're always going to return a strided tensor as output for any kernel.

The situation here is similar to quantized https://github.com/pytorch/pytorch/pull/74560 but unlike with quantization I can't actually delay the choice here. So it's serious trouble for structured kernels design. I'd probably suggest not using structured as a start here.

Comment (cpuhrsch) [reply tools/codegen/dest/register_dispatch_key.py:90]: You want a different layout depending on the dimension. There is only 2D CSR. For a full reduction, we want a 0-dim Tensor, for a partial reduction a 1-dim COO Tensor. That only holds until we introduce dense dimensions for CSR Tensors as well.

**Comment by cpuhrsch** [comment]
Again seeing 
```
/var/lib/jenkins/workspace/build/lib/libtorch_cuda_cu.so: undefined reference to `at::native::sum_csr(at::Tensor const&, c10::optional<c10::ScalarType>)'
collect2: error: ld returned 1 exit status
caffe2/CMakeFiles/cuda_cudnn_test.dir/build.make:111: recipe for target 'bin/cuda_cudnn_test' failed
make[2]: *** [bin/cuda_cudnn_test] Error 1
CMakeFiles/Makefile2:5908: recipe for target 'caffe2/CMakeFiles/cuda_cudnn_test.dir/all' failed
make[1]: *** [caffe2/CMakeFiles/cuda_cudnn_test.dir/all] Error 2
/var/lib/jenkins/workspace/build/lib/libtorch_cuda_cu.so: undefined reference to `at::native::sum_csr(at::Tensor const&, c10::optional<c10::ScalarType>)'
collect2: error: ld returned 1 exit status
test_cpp_c10d/CMakeFiles/example_allreduce.dir/build.make:111: recipe for target 'bin/example_allreduce' failed
make[2]: *** [bin/example_allreduce] Error 1
CMakeFiles/Makefile2:13117: recipe for target 'test_cpp_c10d/CMakeFiles/example_allreduce.dir/all' failed
make[1]: *** [test_cpp_c10d/CMakeFiles/example_allreduce.dir/all] Error 2
```

I'm convinced at this point that this is due to a namespace conflict. SparseCsrTensorMath.cpp imports all of at::sparse

**Comment by cpuhrsch** [comment]
Moving sum_csr to ReduceOps.cpp resolved the linker error after all

**Review by ezyang — APPROVED** [review]
okey dokey
